lib/tcpip/network_wrapper.cc

```diff
@@ -604,30 +611,34 @@ int network_socket_close(Timeout *t, SObj mallocCapability, SObj sealedSocket)
 		  }
 		  // Close the socket.  Another thread will actually clean up the
 		  // memory.
+		  int  ret         = 0;
 		  auto closeStatus = close_socket_retry(t, socket->socket);
 		  if (closeStatus == 0)
 		  {
-			  return -EINVAL;
+			  // The only reason why this would fail is internal
+			  // corruption (did someone already close the
+			  // socket?). Nothing can be done by anyone at this
+			  // stage. Don't return because we would leak
+			  // everything.
+			  ret = -ENOTRECOVERABLE;
 		  }
 		  if (closeStatus != 1)
 		  {
-			  // With some lock, the socket can be freed next time
+			  socket->socketLock.unlock();
+			  // With some luck, the socket can be freed next time
 			  // we try.
 			  return -ETIMEDOUT;
 		  }
 		  socket->socketLock.upgrade_for_destruction();
 		  // Drop the caller's claim on the socket.
-		  int ret = 0;
 		  if (heap_free(mallocCapability, socket->socket) != 0)
 		  {
 			  // This is not supposed to happen, since we did a
 			  // heap_can_free earlier. If it does, we're leaking
 			  // the socket.
 			  Debug::log("Failed to free socket.");
-			  // Release the lock so that we don't leak it.
-			  socket->socketLock.unlock();
 			  // Don't return now, try to at least free the token.
-			  ret = -EINVAL;
+			  ret = -ENOTRECOVERABLE;
```

Comment:
I believe the only way to reach this is to have insufficient stack to call `heap_free`.  Is there any other way that it can happen?

Reply:
Agree. Of course there is also the possibility that the caller frees it concurrently as it owns the heap capability but this has no impact here. And there is also the possibility of a bug somewhere, but in theory it shouldn't happen.

Reply:
We can check for the former case by checking `-ECOMPARTMENTFAIL` on the `heap_free` return and the latter by checking the tag bit.  

In the case where there's a concurrent free, we should continue freeing the other things.

Reply:
> In the case where there's a concurrent free, we should continue freeing the other things.

This is already what we do?

> We can check for the former case by checking -ECOMPARTMENTFAIL on the heap_free return and the latter by checking the tag bit.

What should we check it for exactly? To provide a more explicit error code?